pkg/tsdb/grafana-pyroscope-datasource/instance.go

```diff
@@ -105,17 +107,44 @@ func (d *PyroscopeDatasource) profileTypes(ctx context.Context, req *backend.Cal
 
 func (d *PyroscopeDatasource) labelNames(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
 	ctxLogger := logger.FromContext(ctx)
-	res, err := d.client.LabelNames(ctx)
+
+	u, err := url.Parse(req.URL)
+	if err != nil {
+		ctxLogger.Error("Failed to parse URL", "error", err, "function", logEntrypoint())
+		return err
+	}
+	query := u.Query()
+
+	start, _ := strconv.ParseInt(query.Get("start"), 10, 64)
+	end, _ := strconv.ParseInt(query.Get("end"), 10, 64)
+	labelSelector := query.Get("query")
+
+	labels, err := d.client.LabelNames(ctx, labelSelector, start, end)
 	if err != nil {
 		ctxLogger.Error("Received error from client", "error", err, "function", logEntrypoint())
 		return fmt.Errorf("error calling LabelNames: %v", err)
 	}
-	data, err := json.Marshal(res)
+	matchers, err := parser.ParseMetricSelector(labelSelector)
+	if err != nil {
+		ctxLogger.Error("Could not parse label selector in input", "error", err, "function", logEntrypoint())
+		return fmt.Errorf("error processing label names in response: %v", err)
+	}
+	finalLabels := make([]string, 0)
+Outer:
+	for _, label := range labels {
+		for _, matcher := range matchers {
+			if matcher.Name == label {
+				continue Outer
+			}
+		}
+		finalLabels = append(finalLabels, label)
+	}
```

Comment:
This is a bit ugly, but the gist of it is that we want to remove the labels from the response that are already selected. It could potentially be done in the frontend instead.